src/erf.jl

```diff
@@ -4,40 +4,73 @@ using Base.Math: @horner
 using Base.MPFR: ROUNDING_MODE
 
 for f in (:erf, :erfc)
+    internalf = Symbol(:_, f)
+    libopenlibmf = QuoteNode(f)
+    libopenlibmf0 = QuoteNode(Symbol(f, :f))
+    openspecfunf = QuoteNode(Symbol(:Faddeeva_, f))
+    mpfrf = QuoteNode(Symbol(:mpfr_, f))
     @eval begin
-        ($f)(x::Float64) = ccall(($(string(f)),libopenlibm), Float64, (Float64,), x)
-        ($f)(x::Float32) = ccall(($(string(f,"f")),libopenlibm), Float32, (Float32,), x)
-        ($f)(x::Real) = ($f)(float(x))
-        ($f)(a::Float16) = Float16($f(Float32(a)))
-        ($f)(a::Complex{Float16}) = Complex{Float16}($f(Complex{Float32}(a)))
-        function ($f)(x::BigFloat)
+        $f(x::Number) = $internalf(float(x))
+
+        $internalf(x::Float64) = ccall(($libopenlibmf, libopenlibm), Float64, (Float64,), x)
+        $internalf(x::Float32) = ccall(($libopenlibmf0, libopenlibm), Float32, (Float32,), x)
+        $internalf(x::Float16) = Float16($internalf(Float32(x)))
+
+        $internalf(z::Complex{Float64}) = Complex{Float64}(ccall(($openspecfunf, libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), z, zero(Float64)))
+        $internalf(z::Complex{Float32}) = Complex{Float32}(ccall(($openspecfunf, libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), Complex{Float64}(z), Float64(eps(Float32))))
+        $internalf(z::Complex{Float16}) = Complex{Float16}($internalf(Complex{Float32}(z)))
+
+        function $internalf(x::BigFloat)
             z = BigFloat()
-            ccall(($(string(:mpfr_,f)), :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, Int32), z, x, ROUNDING_MODE[])
+            ccall(($mpfrf, :libmpfr), Int32, (Ref{BigFloat}, Ref{BigFloat}, Int32), z, x, ROUNDING_MODE[])
             return z
         end
-        ($f)(x::AbstractFloat) = error("not implemented for ", typeof(x))
     end
 end
 
-for f in (:erf, :erfc, :erfcx, :erfi, :Dawson)
-    fname = (f === :Dawson) ? :dawson : f
+for f in (:erfcx, :erfi, :dawson)
+    internalf = Symbol(:_, f)
+    openspecfunfsym = Symbol(:Faddeeva_, f === :dawson ? :Dawson : f)
+    openspecfunfF64 = QuoteNode(Symbol(openspecfunfsym, :_re))
+    openspecfunfCF64 = QuoteNode(openspecfunfsym)
     @eval begin
-        ($fname)(z::Complex{Float64}) = Complex{Float64}(ccall(($(string("Faddeeva_",f)),libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), z, zero(Float64)))
-        ($fname)(z::Complex{Float32}) = Complex{Float32}(ccall(($(string("Faddeeva_",f)),libopenspecfun), Complex{Float64}, (Complex{Float64}, Float64), Complex{Float64}(z), Float64(eps(Float32))))
+        $f(x::Number) = $internalf(float(x))
+
+        $internalf(x::Float64) = ccall(($openspecfunfF64, libopenspecfun), Float64, (Float64,), x)
+        $internalf(x::Float32) = Float32($internalf(Float64(x)))
+        $internalf(x::Float16) = Float16($internalf(Float64(x)))
```

Comment:
This adds support for `Float16` in a similar way as already present for `erf` etc. above. If desired I can move it to a separate PR but it seemed like it was missing accidentally.